### PR TITLE
Runtime permission support for Android 23+ Compatibility

### PR DESCRIPTION
**Pull request by OptrixAU**
This resolves the permission issue originally discussed in issue #78 and #84 by adding runtime permission checks and requests to the Activity. This will allow ARToolkit apps to run on Samsung Galaxy S7 phones and above. 

It doesn't currently have a graceful failure mode if someone refuses to allow the application to access the camera, and the updates are only available for the **Android Studio** version of the project, not the Eclipse one.

There is also one additional change - a public property **IsRunning** on the ARController class that can be used to tell when the camera is ready. I frequently need to patch this in by hand, as I need to know when to remove my 'Please Wait' UI, and I'd imagine that it's an _extremely_ common requirement.

### CODE REVIEW

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/CameraSurface.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ public void surfaceCreated(SurfaceHolder holder) {
                 camera = Camera.open();
             }
         } catch (RuntimeException exception) {
-            Log.e(TAG, "Cannot open camera. It may be in use by another process.");
+			Log.e(TAG, "Cannot Open Camera - " + exception.getMessage());
```
Comment (BoiseBaked) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/CameraSurface.java]: I like.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java [hunk 2]
```diff
@@ -114,6 +120,33 @@ protected void onCreate(Bundle savedInstanceState) {
         int resID = getResources().getIdentifier("preferences", "xml", getPackageName());
         PreferenceManager.setDefaultValues(this, resID, false);
 
+		//Request permission to use the camera on android 23+
+        int permissionCheck = ContextCompat.checkSelfPermission(this.getApplicationContext(), Manifest.permission.CAMERA);
+        if (permissionCheck != PackageManager.PERMISSION_GRANTED)
```
Comment (BoiseBaked) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: I think just after this if,  if (shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) is required to execute a toast explaining why the permissions is required. After that the requestPermissions() interface is called.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java [hunk 3]
```diff
@@ -114,6 +120,33 @@ protected void onCreate(Bundle savedInstanceState) {
...
+	//Handle the result of asking the user for camera permission.
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           String permissions[], int[] grantResults) {
+        switch (requestCode) {
+            case 77: {
+                // If request is cancelled, the result arrays are empty.
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+
+                    //Camera permission granted!
```
Comment (BoiseBaked) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: No state to set I suppose, camera should just work hereafter. Anyways, make make a toast stating access to permission to camera granted by the user.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java (hunk 2 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: It's not actually REQUIRED to do so (the API specification mentions that you don't really need to do this if the reason for the permission is obvious).

Additionally, if you popped up a toast, you'd be doing the whole thing asynchronously (ie. your creation code would finish before you've got your permission). You'd need to wait around for the permission response and once it arrives re-create the camera objects.

By doing it inline and without explaining yourself, you don't have to deal with the asynchronous callback problems - but I do understand if you decide it's worth going the async route.

File: src/Unity/Assets/ARToolKit5-Unity/Scripts/ARController.cs [hunk 4]
```diff
@@ -86,6 +86,9 @@ public class ARController : MonoBehaviour
     public bool QuitOnEscOrBack = true;
     public bool AutoStartAR = true;
 
+	[HideInInspector]
+	public bool IsRunning { get { return _running; } }
```
Comment (BoiseBaked) [inline src/Unity/Assets/ARToolKit5-Unity/Scripts/ARController.cs]: Where is this called or used? May want to add a comment indicating that.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java (hunk 3 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: You can feel free to add that - it's certainly a good idea if you're going the asynchronous route. Certainly a toast saying that camera access is required if you're silly enough to DENY the permission is a great idea.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java [hunk 5]
```diff
@@ -114,6 +120,33 @@ protected void onCreate(Bundle savedInstanceState) {
         int resID = getResources().getIdentifier("preferences", "xml", getPackageName());
         PreferenceManager.setDefaultValues(this, resID, false);
 
+		//Request permission to use the camera on android 23+
+        int permissionCheck = ContextCompat.checkSelfPermission(this.getApplicationContext(), Manifest.permission.CAMERA);
+        if (permissionCheck != PackageManager.PERMISSION_GRANTED)
+        {
+            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, 77);
```
Comment (BoiseBaked) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: Create a int constant for the value of 77.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java [hunk 6]
```diff
@@ -114,6 +120,33 @@ protected void onCreate(Bundle savedInstanceState) {
...
+        switch (requestCode) {
+            case 77: {
+                // If request is cancelled, the result arrays are empty.
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+
+                    //Camera permission granted!
+
+                } else {
+
+                    // TODO: Fail gracefully here.
```
Comment (BoiseBaked) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: Maybe a toast here explaining permission was not granted. Also, let's find a way to gracefully terminate the app. This is always hard to do given that there is no real terminate state other than an abnormal termination, i.e. a crash. I guess reset state so onResume can start from the beginning.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/build.gradle [hunk 7]
```diff
@@ -16,12 +16,17 @@ android {
...
+        sourceCompatibility JavaVersion.VERSION_1_7
+        targetCompatibility JavaVersion.VERSION_1_7
+    }
 }
 
 dependencies {
     compile project(':aRBaseLib-release')
     compile project(':classes')
     compile project(':BT200Ctrl')
+    compile 'com.android.support:support-v4:23.0.0'
```
Comment (BoiseBaked) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/build.gradle]: I'm implementing this on other Android projects in the ARToolKit 5 repo and in the build.gradle file dependencies in all the examples I've seen, the latest and greatest library for compileSDKVersion 23 is compile 'com.android.support:appcompat-v7:23.3.0' to defined all the appcompat dependencies. Why are your changes building without the dependency?

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java (hunk 5 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: Fair point - no magic numbers.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/build.gradle (hunk 7 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/build.gradle]: I just used the first support library to come to hand (it was also the one used in the sample documentation on how to use the new permission request system).

As there was no existing support library being used in the Unity library, I linked to the v4 version. There would be no issue moving to the v7 (and tweaking the revision number) if that's the standard used in ARToolkit.

Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/build.gradle]: Updated it to 23.3.0 in the latest commit, but still using the support-v4 library (I'm getting crashes on the v7 and have insufficient time to diagnose the problem at the moment).

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java (hunk 6 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: I offer a toast, then we are left with a blank screen until they close and restart. Further logic might be nice, but not critical.

File: src/Unity/Assets/ARToolKit5-Unity/Scripts/ARController.cs (hunk 4 above)
Comment (OptrixAU) [inline src/Unity/Assets/ARToolKit5-Unity/Scripts/ARController.cs]: Further comments added.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/CameraSurface.java (hunk 1 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/CameraSurface.java]: Should be resolved in latest commit.

File: src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java (hunk 2 above)
Comment (OptrixAU) [inline src/Android_Unity_Player_Source/AndroidStudioProj/UnityARPlayer/src/main/java/org/artoolkit/ar/unity/UnityARPlayerActivity.java]: This is **not** resolved in the latest commit, as while I understand your perspective, I believe it to be unnecessary. 

I've got no valid arguments _against_  doing it, just none _for_ it either. I certainly won't complain if someone else patches it in.

**Pull request merged**